### PR TITLE
*: support ssl/tls

### PR DESCRIPTION
**Pull request by Connor1996**
As mentioned in #865 
pd-ctl and pd-client can connect with pd in ssl/tls connection, and etcd-client with etcd-server too.
when tls-ca-path in config.toml is set, the connection among etcd peers will also use ssl/tls with auto generated cert and key file.

note: relative urls should use https schema, or it will still not use ssl/tls

when using ssl/tls, there is a bug of embed etcd that will cause panic when stopping pd process.
https://github.com/coreos/etcd/issues/8916

### CODE REVIEW

File: pd-client/client.go [hunk 1]
```diff
@@ -93,10 +98,14 @@ type client struct {
...
 	cancel context.CancelFunc
+
+	tlsCAPath   string
+	tlsCertPath string
+	tlsKeyPath  string
 }
 
 // NewClient creates a PD client.
-func NewClient(pdAddrs []string) (Client, error) {
+func NewClient(pdAddrs []string, tlsCAPath, tlsCertPath, tlsKeyPath string) (Client, error) {
```
Comment (siddontang) [inline pd-client/client.go]: move these to a SecurityOption struct.

**Comment by siddontang** [comment]
I prefer to follow https://coreos.com/etcd/docs/latest/op-guide/security.html and use the same name flag. 

Btw, I don't think we need to support auto TLS now.

/cc @disksing

File: conf/config.toml [hunk 2]
```diff
@@ -17,6 +17,19 @@ initial-cluster-state = "new"
...
 tso-save-interval = "3s"
 
+# Path of file that contains list of trusted SSL CAs.
+tls-ca-path = ""
+# Path of file that contains X509 certificate in PEM format.
+tls-cert-path = ""
+# Path of file that contains X509 key in PEM format.
+tls-key-path = ""
+# enable client certificate auth, if true following two settings shouldn't be empty
+client-cert-auth = false
```
Comment (disksing) [inline conf/config.toml]: As follower pd-servers redirect all requests to leader, so typically if if TLS is enabled on server side, client side must enable it too. Therefore I think this flag looks redundant to me.

Comment (siddontang) [reply conf/config.toml]: @disksing @huachaohuang 

Maybe we should let the client redirect to the new leader to reduce the latency.

File: pd-client/client.go [hunk 3]
```diff
@@ -213,11 +228,46 @@ func (c *client) getOrCreateGRPCConn(addr string) (*grpc.ClientConn, error) {
...
+
+		// Create a certificate pool from the certificate authority
+		certPool := x509.NewCertPool()
+		ca, err := ioutil.ReadFile(c.security.TLSCAPath)
+		if err != nil {
+			return nil, errors.Errorf("could not read ca certificate: %s", err)
+		}
+
+		// Append the certificates from the CA
+		if ok := certPool.AppendCertsFromPEM(ca); !ok {
```
Comment (disksing) [inline pd-client/client.go]: `if !certPool.AppendCertsFromPEM(ca) {`

File: pdctl/command/global.go [hunk 4]
```diff
@@ -23,20 +25,52 @@ import (
...
+
+	// Create a certificate pool from the certificate authority
+	certPool := x509.NewCertPool()
+	ca, err := ioutil.ReadFile(tlsCAPath)
+	if err != nil {
+		return errors.Errorf("could not read ca certificate: %s", err)
+	}
+
+	// Append the certificates from the CA
+	if ok := certPool.AppendCertsFromPEM(ca); !ok {
```
Comment (disksing) [inline pdctl/command/global.go]: `if !certPool.AppendCertsFromPEM(ca) {`

**Comment by disksing** [comment]
LGTM. Have you tested it with tikv-server?

**Comment by Connor1996** [comment]
@disksing Yes

**Comment by siddontang** [comment]
LGTM

File: server/config.go [hunk 5]
```diff
@@ -85,6 +85,10 @@ type Config struct {
 	// ElectionInterval is the interval for etcd Raft election.
 	ElectionInterval typeutil.Duration `toml:"election-interval"`
 
+	TLSCAPath   string `toml:"tls-cacert-path" json:"tls-cacert-path"`
```
Comment (siddontang) [inline server/config.go]: I think we can remove tls and use a security section like TiKV does

Comment (siddontang) [reply server/config.go]: ping @Connor1996

**Comment by Connor1996** [comment]
PTAL @siddontang @disksing

**Comment by Connor1996** [comment]
Fix #865

**Pull request merged**